### PR TITLE
21.3 fb ios push notification changes

### PR DESCRIPTION
**Pull request by rohitn-boston**
#### Rationale
Push Notification issue for the iOS application is resolved.

#### Related Pull Requests
* N/A

#### Changes
* Spanish language support added

### CODE REVIEW

**Comment by labkey-adam** [comment]
@rohitn-boston I'll review this PR more thoroughly tomorrow, but at first glance, it looks like you've added a SQL script that will never run on existing servers since you haven't bumped the module version. The module version is now 21.000, so I believe you want to name the script `fdahpuserregws-21.000-21.001.sql` and bump the module version to 21.001. You'll also need to update `fdahpuserregws.xml` to match the changes in the script. And you'll need to update jars.txt to accommodate the jar file change (see the credits page). You'll see that several tests are failing and pointing out these problems. There are also a few tests that are running that shouldn't be, like the StudyShortTest; you can ignore that. Let me know if you have questions.

**Comment by rohitn-boston** [comment]
@labkey-adam We have made the changes except the one to update jars.txt. This change we are not familiar with. Can you help us with updating the file? The apns jar file is removed and added as a dependency. I am not sure about the changes that we have to make in jars.txt in this case.

**Comment by labkey-adam** [comment]
@rohitn-boston The built-in credits page documents all the primary jar file dependencies that ship with a distribution. This ensures that dependencies are documented and credited (which is sometimes required by the licenses). It also helps us audit the licenses being used. You can visit the credits page of any LabKey Server by visiting the Admin Console -> Credits link. Or you can navigate directly to this page (no permissions required); for example, see https://www.labkey.org/admin-credits.view. The jars.txt files included in every module provides documentation for its jar files and the page verifies that the jar files in the module match the documentation. `BasicTest.testCredits` then verifies that no warnings appear on the credits page; that test is failing. You'll see the warning clearly if you visit the credits page on your development instance. You can also see the failure on TeamCity; visit the test failure page (e.g., https://teamcity.labkey.org/viewLog.html?buildId=1429094&buildTypeId=LabKey_213Release_FdaMyStudies_MyStudiesRegistrationPostgres), click "Artifacts," expand "BasicTest.tar.gz", and click the HTML artifact. You'll see:

**WARNING: The following JAR file was found in your webapp but is not documented in the FdahpUserRegWS Module jars.txt:**
pushy-0.14.1.jar

**WARNING: The following JAR file is documented in the FdahpUserRegWS Module jars.txt but was not found in your webapp:**
apns-1.0.0.Beta6.jar

In your jars.txt file, replace the reference to apns with the details about pushy and this warning will disappear.

**Review by labkey-tchad — CHANGES_REQUESTED** [review]
Please merge `release21.3-SNAPSHOT` into this feature branch to pull in the regression tests added in #8

**Comment by praveenb-btc** [comment]
> Please merge `release21.3-SNAPSHOT` into this feature branch to pull in the regression tests added in #8
@labkey-tchad 
After merging  release21.3-SNAPSHOT i am getting standalone plugin not found please find the attachment for reference
 
Screenshot (34)

**Comment by praveenb-btc** [comment]
> @rohitn-boston The built-in credits page documents all the primary jar file dependencies that ship with a distribution. This ensures that dependencies are documented and credited (which is sometimes required by the licenses). It also helps us audit the licenses being used. You can visit the credits page of any LabKey Server by visiting the Admin Console -> Credits link. Or you can navigate directly to this page (no permissions required); for example, see https://www.labkey.org/admin-credits.view. The jars.txt files included in every module provides documentation for its jar files and the page verifies that the jar files in the module match the documentation. `BasicTest.testCredits` then verifies that no warnings appear on the credits page; that test is failing. You'll see the warning clearly if you visit the credits page on your development instance. You can also see the failure on TeamCity; visit the test failure page (e.g., https://teamcity.labkey.org/viewLog.html?buildId=1429094&buildTypeId=LabKey_213Release_FdaMyStudies_MyStudiesRegistrationPostgres), click "Artifacts," expand "BasicTest.tar.gz", and click the HTML artifact. You'll see:
> 
> **WARNING: The following JAR file was found in your webapp but is not documented in the FdahpUserRegWS Module jars.txt:**
> pushy-0.14.1.jar
> 
> **WARNING: The following JAR file is documented in the FdahpUserRegWS Module jars.txt but was not found in your webapp:**
> apns-1.0.0.Beta6.jar
> 
> In your jars.txt file, replace the reference to apns with the details about pushy and this warning will disappear.

Even after replacing apns with pushy in jars.txt file Test is still failing please check this commit [29b88ad](https://github.com/FDA-MyStudies/UserReg-WS/pull/14/commits/29b88ad94a9b080ac2d74bd2aed68edb9bdeb538)

**Comment by rohitn-boston** [comment]
> > @rohitn-boston The built-in credits page documents all the primary jar file dependencies that ship with a distribution. This ensures that dependencies are documented and credited (which is sometimes required by the licenses). It also helps us audit the licenses being used. You can visit the credits page of any LabKey Server by visiting the Admin Console -> Credits link. Or you can navigate directly to this page (no permissions required); for example, see https://www.labkey.org/admin-credits.view. The jars.txt files included in every module provides documentation for its jar files and the page verifies that the jar files in the module match the documentation. `BasicTest.testCredits` then verifies that no warnings appear on the credits page; that test is failing. You'll see the warning clearly if you visit the credits page on your development instance. You can also see the failure on TeamCity; visit the test failure page (e.g., https://teamcity.labkey.org/viewLog.html?buildId=1429094&buildTypeId=LabKey_213Release_FdaMyStudies_MyStudiesRegistrationPostgres), click "Artifacts," expand "BasicTest.tar.gz", and click the HTML artifact. You'll see:
> > **WARNING: The following JAR file was found in your webapp but is not documented in the FdahpUserRegWS Module jars.txt:**
> > pushy-0.14.1.jar
> > **WARNING: The following JAR file is documented in the FdahpUserRegWS Module jars.txt but was not found in your webapp:**
> > apns-1.0.0.Beta6.jar
> > In your jars.txt file, replace the reference to apns with the details about pushy and this warning will disappear.
> 
> Even after replacing apns with pushy in jars.txt file Test is still failing please check this commit [29b88ad](https://github.com/FDA-MyStudies/UserReg-WS/pull/14/commits/29b88ad94a9b080ac2d74bd2aed68edb9bdeb538)

@labkey-adam Can you provide some insights on this?

**Comment by labkey-adam** [comment]
> > > @rohitn-boston The built-in credits page documents all the primary jar file dependencies that ship with a distribution. This ensures that dependencies are documented and credited (which is sometimes required by the licenses). It also helps us audit the licenses being used. You can visit the credits page of any LabKey Server by visiting the Admin Console -> Credits link. Or you can navigate directly to this page (no permissions required); for example, see https://www.labkey.org/admin-credits.view. The jars.txt files included in every module provides documentation for its jar files and the page verifies that the jar files in the module match the documentation. `BasicTest.testCredits` then verifies that no warnings appear on the credits page; that test is failing. You'll see the warning clearly if you visit the credits page on your development instance. You can also see the failure on TeamCity; visit the test failure page (e.g., https://teamcity.labkey.org/viewLog.html?buildId=1429094&buildTypeId=LabKey_213Release_FdaMyStudies_MyStudiesRegistrationPostgres), click "Artifacts," expand "BasicTest.tar.gz", and click the HTML artifact. You'll see:
> > > **WARNING: The following JAR file was found in your webapp but is not documented in the FdahpUserRegWS Module jars.txt:**
> > > pushy-0.14.1.jar
> > > **WARNING: The following JAR file is documented in the FdahpUserRegWS Module jars.txt but was not found in your webapp:**
> > > apns-1.0.0.Beta6.jar
> > > In your jars.txt file, replace the reference to apns with the details about pushy and this warning will disappear.
> > 
> > 
> > Even after replacing apns with pushy in jars.txt file Test is still failing please check this commit [29b88ad](https://github.com/FDA-MyStudies/UserReg-WS/pull/14/commits/29b88ad94a9b080ac2d74bd2aed68edb9bdeb538)
> 
> @labkey-adam Can you provide some insights on this?

You have a typo in jars.txt. The jar name is `pushy-0.14.1.jar` but you have a colon (:) instead of a dash (-). The names must match exactly. If you visit the HTML artifact and scroll down you'll see the mismatch clearly: https://teamcity.labkey.org/repository/download/LabKey_213Release_FdaMyStudies_MyStudiesRegistrationPostgres/1430151:id/BasicTest.tar.gz!/202106030347BasicTest%23testCredits.html

File: resources/credits/jars.txt [hunk 1]
```diff
@@ -1,5 +1,5 @@
 {table}
 Filename|Component|Version|Source|License|Purpose
-apns-1.0.0.Beta6.jar|apns|1.0.0.Beta6|{link:java-apns|https://notnoop.github.io/java-apns/}|{link:BSD|https://github.com/notnoop/java-apns/blob/master/LICENSE}|Java Apple Push Notification Service Library
+pushy:0.14.1.jar|pushy|0.14.1|{link:pushy|https://pushy-apns.org/}|{link:MIT|https://github.com/jchambers/pushy/blob/master/LICENSE.md}|Java Apple Push Notification Service Library
```
Comment (labkey-tchad) [inline resources/credits/jars.txt]: This needs to match the file name exactly.
```suggestion
pushy-0.14.1.jar|pushy|0.14.1|{link:pushy|https://pushy-apns.org/}|{link:MIT|https://github.com/jchambers/pushy/blob/master/LICENSE.md}|Java Apple Push Notification Service Library
```

**Comment by labkey-tchad** [comment]
> > Please merge `release21.3-SNAPSHOT` into this feature branch to pull in the regression tests added in #8
> > @labkey-tchad
> > After merging  release21.3-SNAPSHOT i am getting standalone plugin not found please find the attachment for reference
> 
> `screenshot`

That error can be resolved by updating your root labkey enlistment (`D:\FDA-Project\labkeyHome-21.3`). That will pull in the change that updates the gradle plugin.

**Review by labkey-adam — APPROVED** [review]

**Comment by labkey-adam** [comment]
Changes look reasonable and all tests are now passing! @labkey-tchad requested a change that you made; you should always "re-request review" from that reviewer so they know the change has been made. I clicked the icon in this case, but good to know for the future.

**Review by labkey-tchad — APPROVED** [review]

**Comment by labkey-tchad** [comment]
@rohitn-boston Please use ["Squash and merge"](https://docs.github.com/en/github/collaborating-with-pull-requests/incorporating-changes-from-a-pull-request/merging-a-pull-request#merging-a-pull-request-on-github) for this PR to keep the history clean.

**Comment by rohitn-boston** [comment]
> Changes look reasonable and all tests are now passing! @labkey-tchad requested a change that you made; you should always "re-request review" from that reviewer so they know the change has been made. I clicked the icon in this case, but good to know for the future.

We will do it going forward.

**Pull request merged**